File: skip-the-dishes-query-event-handler/src/main/java/br/com/skip/the/dishes/query/event/handler/order/OrderEventHandler.java
package br.com.skip.the.dishes.query.event.handler.order;

import br.com.skip.the.dishes.domain.order.DeliveryAddress;
import br.com.skip.the.dishes.domain.order.commons.OrderApplier;
import br.com.skip.the.dishes.domain.order.commons.OrderEvent;
import br.com.skip.the.dishes.domain.order.events.*;
import br.com.skip.the.dishes.query.repository.order.OrderEntity;
import br.com.skip.the.dishes.query.repository.order.OrderRepository;
import br.com.skip.the.dishes.query.repository.product.ProductEntity;
import br.com.skip.the.dishes.query.repository.product.ProductRepository;
import br.com.zup.eventsourcing.core.*;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
public class OrderEventHandler implements EventHandler {

    private OrderApplier applier = new Applier();

    private OrderRepository orderRepository;

    private ProductRepository productRepository;

    public OrderEventHandler(OrderRepository orderRepository, ProductRepository productRepository) {
        this.orderRepository = orderRepository;
        this.productRepository = productRepository;
    }

    @Override
    public void handle(AggregateId aggregateId, Event event, MetaData metaData, AggregateVersion aggregateVersion) {
        ((OrderEvent)event).accept(aggregateId, applier);
    }

    private class Applier implements OrderApplier {
        @Override
        public void on(AggregateId aggregateId, OrderCreated orderCreated) {
            OrderEntity orderEntity = new OrderEntity(aggregateId.getValue(), orderCreated.getCustomerId());
            orderEntity.setStatus(orderCreated.getStatus().toString());
            orderRepository.save(orderEntity);
        }

        @Override
        public void on(AggregateId aggregateId, ProductAdded productAdded) {
            OrderEntity orderEntity = orderRepository.findById(aggregateId.getValue()).get();
            ProductEntity productEntity = productRepository.findById(productAdded.getProductId()).get();

            orderEntity.getProducts().add(productEntity);

            orderRepository.save(orderEntity);
        }

        @Override
        public void on(AggregateId aggregateId, ProductDeleted productDeleted) {
            OrderEntity orderEntity = orderRepository.findById(aggregateId.getValue()).get();
            ProductEntity productEntity = productRepository.findById(productDeleted.getProductId()).get();

            orderEntity.getProducts().remove(productEntity);

            orderRepository.save(orderEntity);
        }

        @Override
        public void on(AggregateId aggregateId, OrderCancelled orderCancelled) {
            OrderEntity orderEntity = orderRepository.findById(aggregateId.getValue()).get();
            orderEntity.setStatus(orderCancelled.getOrderStatus().toString());
            orderRepository.save(orderEntity);
        }

        @Override
        public void on(AggregateId aggregateId, OrderRequested orderRequested) {
            OrderEntity orderEntity = orderRepository.findById(aggregateId.getValue()).get();
            orderEntity.setStatus(orderRequested.getOrderStatus().toString());
            orderRepository.save(orderEntity);
        }

        @Override
        public void on(AggregateId aggregateId, DeliveryAddressUpdated deliveryAddressUpdated) {
            OrderEntity orderEntity = orderRepository.findById(aggregateId.getValue()).get();
            DeliveryAddress deliveryAddress = deliveryAddressUpdated.getDeliveryAddress();
            orderEntity.setDeliveryAddress(deliveryAddress.getDeliveryAddress());
            orderEntity.setContact(deliveryAddress.getContact());
            orderRepository.save(orderEntity);
        }
    }
}

File: skip-the-dishes-query-event-handler/src/main/java/br/com/skip/the/dishes/query/event/handler/order/OrderEventHandler.kt
package br.com.skip.the.dishes.query.event.handler.order

import br.com.skip.the.dishes.domain.order.elements.*
import br.com.skip.the.dishes.query.repository.order.OrderEntity
import br.com.skip.the.dishes.query.repository.order.OrderRepository
import br.com.skip.the.dishes.query.repository.product.ProductRepository
import br.com.zup.eventsourcing.core.AggregateId
import br.com.zup.eventsourcing.core.AggregateVersion
import br.com.zup.eventsourcing.core.Event
import br.com.zup.eventsourcing.core.EventHandler
import br.com.zup.eventsourcing.core.MetaData
import org.springframework.stereotype.Component

@Component
class OrderEventHandler(private val orderRepository: OrderRepository, private val productRepository: ProductRepository) : EventHandler, Handler {

    override fun handle(aggregateId: AggregateId, event: Event, metaData: MetaData, version: AggregateVersion) {
        (event as OrderEvent).apply(aggregateId, this)
    }

    override fun on(orderCreated: OrderCreated) {
        val orderEntity = OrderEntity(
                id = orderCreated.orderId,
                customerId = orderCreated.customerId,
                status = orderCreated.status.toString()
        )
        orderRepository.save(orderEntity)
    }

    override fun on(aggregateId: AggregateId, productAdded: ProductAdded) {
        val productEntity = productRepository.findById(productAdded.productId).get()

        orderRepository
                .findById(aggregateId.value)
                .get()
                .let { it.copy(products = it.products + productEntity) }
                .also { orderRepository.save(it) }
    }

    override fun on(aggregateId: AggregateId, productDeleted: ProductDeleted) {
        val productEntity = productRepository.findById(productDeleted.productId).get()

        orderRepository
                .findById(aggregateId.value)
                .get()
                .let { it.copy(products = it.products.filter { it.id != productEntity.id }) }
                .also { orderRepository.save(it) }
    }

    override fun on(aggregateId: AggregateId, orderCancelled: OrderCancelled) {
        orderRepository
                .findById(aggregateId.value)
                .get()
                .let { it.copy(status = orderCancelled.orderStatus.toString()) }
                .also { orderRepository.save(it) }
    }

    override fun on(aggregateId: AggregateId, orderRequested: OrderRequested) {
        orderRepository
                .findById(aggregateId.value)
                .get()
                .let { it.copy(status = orderRequested.orderStatus.toString()) }
                .also { orderRepository.save(it) }
    }

    override fun on(aggregateId: AggregateId, deliveryAddressUpdated: DeliveryAddressUpdated) {
        orderRepository.findById(aggregateId.value).get()
                .let {
                    it.copy(
                            deliveryAddress = deliveryAddressUpdated.deliveryAddress.deliveryAddress,
                            contact = deliveryAddressUpdated.deliveryAddress.contact
                    )
                }
                .also { orderRepository.save(it) }
    }
}

